### PR TITLE
[regression-test](Variant) add compaction case for variant and fix bugs

### PR DESCRIPTION
**Pull request by eldenmoon**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by eldenmoon** [comment]
run buildall

File: be/src/vec/json/path_in_data.cpp [hunk 1]
```diff
@@ -151,10 +151,17 @@ size_t PathInData::Hash::operator()(const PathInData& value) const {
...
+
+PathInData PathInData::pop_nfront(size_t n) const {
+    if (n >= parts.size()) {
+        return {};
+    }
     PathInData new_path;
     Parts new_parts;
     if (!parts.empty()) {
-        std::copy(parts.begin() + 1, parts.end(), std::back_inserter(new_parts));
+        std::copy(parts.begin() + n, parts.end(), std::back_inserter(new_parts));
```
Comment (xiaokang) [inline be/src/vec/json/path_in_data.cpp:164]: Does 'pop_nfront' remove n front element using std::copy?

File: be/src/vec/functions/function_cast.h [hunk 2]
```diff
@@ -2013,8 +2013,16 @@ class FunctionCast final : public IFunctionBase {
...
-            if (variant.is_scalar_variant()) {
+            // It's important to convert as many elements as possible in this context. For instance,
+            // if the root of this variant column is a number column, converting it to a number column
+            // is acceptable. However, if the destination type is a string and root is none scalar root, then
+            // we should convert the entire tree to a string.
+            bool is_root_valuable =
+                    variant.is_scalar_variant() ||
+                    (!variant.is_null_root() &&
+                     !WhichDataType(remove_nullable(variant.get_root_type())).is_nothing() &&
+                     !WhichDataType(data_type_to).is_string());
```
Comment (xiaokang) [inline be/src/vec/functions/function_cast.h:2024]: data_type_to is not string?

File: be/src/vec/json/path_in_data.cpp [hunk 3]
```diff
@@ -151,10 +151,17 @@ size_t PathInData::Hash::operator()(const PathInData& value) const {
 }
 
 PathInData PathInData::pop_front() const {
+    return pop_nfront(1);
+}
+
+PathInData PathInData::pop_nfront(size_t n) const {
```
Comment (xiaokang) [inline be/src/vec/json/path_in_data.cpp:157]: pop_nfront is only used once, why change it?

File: be/src/olap/rowset/segment_v2/hierarchical_data_reader.cpp [hunk 4]
```diff
@@ -32,11 +32,12 @@ namespace doris {
 namespace segment_v2 {
 
 Status HierarchicalDataReader::create(std::unique_ptr<ColumnIterator>* reader,
+                                      vectorized::PathInData path,
                                       const SubcolumnColumnReaders::Node* node,
                                       const SubcolumnColumnReaders::Node* root,
                                       bool output_as_raw_json) {
     // None leave node need merge with root
-    auto* stream_iter = new HierarchicalDataReader(node->path, output_as_raw_json);
+    auto* stream_iter = new HierarchicalDataReader(path, output_as_raw_json);
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/hierarchical_data_reader.cpp:40]: what's the difference?

**Review by xiaokang — DISMISSED** [review]
LGTM

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Review by xiaokang — APPROVED** [review]
LGTM

**Review by airborne12 — APPROVED** [review]
LGTM

**Pull request merged**